cmd/tar2ext4/tar2ext4.go

```diff
@@ -1,63 +1,317 @@
 package main
 
 import (
-	"flag"
+	"context"
 	"fmt"
 	"io"
 	"os"
 
 	"github.com/Microsoft/hcsshim/ext4/tar2ext4"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/urfave/cli"
 )
 
 var (
-	input      = flag.String("i", "", "input file")
-	output     = flag.String("o", "", "output file")
-	overlay    = flag.Bool("overlay", false, "produce overlayfs-compatible layer image")
-	vhd        = flag.Bool("vhd", false, "add a VHD footer to the end of the image")
-	inlineData = flag.Bool("inline", false, "write small file data into the inode; not compatible with DAX")
+	// tar2ext4 convert flags
+	outputFlag         = "output"
+	overlayFlag        = "overlay"
+	vhdFlag            = "vhd"
+	inlineFlag         = "inline"
+	partitionGUIDsFlag = "guid"
+	diskGUIDFlag       = "disk-guid"
+	gptFlag            = "gpt"
+
+	// common flags
+	inputFlag = "input"
+
+	// flags for reading gpt disks
+	pmbrFlag               = "pmbr"
+	headerFlag             = "header"
+	altHeaderFlag          = "alt-header"
+	entryFlag              = "entry"
+	altEntryFlag           = "alt-entry"
+	readAllBytesFlag       = "all"
+	partitionFlag          = "partition"
+	paritionSuperblockFlag = "partition-sb"
+	vhdFooterFlag          = "vhd-footer"
 )
 
 func main() {
-	flag.Parse()
-	if flag.NArg() != 0 || len(*output) == 0 {
-		flag.Usage()
-		os.Exit(1)
+	app := cli.NewApp()
+	app.Name = "tar2ext4"
+	app.Usage = "converts tar file(s) into vhd(s)"
+	app.Commands = []cli.Command{
+		readGPTCommand,
 	}
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:  outputFlag + ",o",
+			Usage: "output file",
+		},
+		cli.BoolFlag{
+			Name:  overlayFlag,
+			Usage: "produce overlayfs-compatible layer image",
+		},
+		cli.BoolFlag{
+			Name:  vhdFlag,
+			Usage: "add a VHD footer to the end of the image",
+		},
+		cli.BoolFlag{
+			Name:  inlineFlag,
+			Usage: "write small file data into the inode; not compatible with DAX",
+		},
+		cli.StringSliceFlag{
+			Name:  inputFlag + ",i",
+			Usage: "input file(s)",
+		},
+		cli.StringSliceFlag{
+			Name:  partitionGUIDsFlag + ",g",
+			Usage: "guid(s) to use for gpt partitions",
+		},
+		cli.StringFlag{
+			Name:  diskGUIDFlag,
+			Usage: "guid for the resulting gpt disk",
+		},
+		cli.BoolFlag{
+			Name:  gptFlag,
+			Usage: "indicates if this disk should be produced as a gpt disk",
+		},
+	}
+	app.Action = func(cliCtx *cli.Context) (err error) {
```

Comment:
minor nit: this function could be defined outside of `main` to reduce the total levels of indentation and have main read cleaner

Reply:
Keeping this here since I moved the gpt code to the new cmd/images tool